### PR TITLE
[GH-32] Add command autocomplete

### LINKED ISSUE #32: Add slash command autocomplete functionality
From 5.24 MM server has a slash command autocomplete functionality feature. An API can be found [here](https://github.com/mattermost/mattermost-server/blob/master/model/command_autocomplete.go#L28)
To implement the feature add `AutocompleteData` struct to the `Command` when registering a slash command.

Similar PRs:
For Todo plugin https://github.com/mattermost/mattermost-plugin-todo/pull/69
For Agenda plugin https://github.com/mattermost/mattermost-plugin-agenda/pull/40

### PR DESCRIPTION
**Pull request by nathanaelhoun**
#### Summary

- [x] Add slash command autocomplete
Peek 2020-10-05 23-30
(the commands are now in alphabetical order in the list)

#### Ticket Link

Closes #32

### CODE REVIEW

**Review by aaronrothschild — APPROVED** [review]
Looks good to me @nathanaelhoun thanks for your contribution - this will certainly make the command easier to use!

File: server/command.go [hunk 1]
```diff
@@ -49,9 +49,73 @@ func getCommand() *model.Command {
...
 		AutoCompleteHint: "[command]",
+		AutocompleteData: getAutocompleteData(),
 	}
 }
 
+func getAutocompleteData() *model.AutocompleteData {
+	jenkins := model.NewAutocompleteData("jenkins", "[subcommand]", "A Mattermost plugin to interact with Jenkins")
+
+	connect := model.NewAutocompleteData("connect", "[username] [APIToken]", "Connect to your Jenkins account")
+	connect.AddTextArgument("Your username of Jenkins", "[username]", "")
```
Comment (mickmister) [inline server/command.go:60]:
```suggestion
	connect.AddTextArgument("The username of your Jenkins account", "[username]", "")
```

File: server/command.go [hunk 2]
```diff
@@ -49,9 +49,73 @@ func getCommand() *model.Command {
...
+	connect.AddTextArgument("You API token from Jenkins", "[API token]", "")
```
Comment (mickmister) [inline server/command.go:61]:
```suggestion
	connect.AddTextArgument("Your API token from Jenkins", "[API token]", "")
```

File: server/command.go [hunk 3]
```diff
@@ -49,9 +49,73 @@ func getCommand() *model.Command {
...
+
+	disconnect := model.NewAutocompleteData("disconnect", "", "Disconnect from your Jenkins account")
+
+	createjob := model.NewAutocompleteData("createjob", "", "Create a Jenkins job using contents of config.xml")
```
Comment (mickmister) [inline server/command.go:65]:
```suggestion
	createjob := model.NewAutocompleteData("createjob", "", "Create a Jenkins job using the contents of a config.xml file")
```

File: server/command.go [hunk 4]
```diff
@@ -49,9 +49,73 @@ func getCommand() *model.Command {
...
 		AutoCompleteDesc: "Available commands: connect, disconnect, me, build, get-artifacts, test-results, get-log, abort, disable, enable, delete, safe-restart, plugins, createjob, help",
 		AutoCompleteHint: "[command]",
+		AutocompleteData: getAutocompleteData(),
 	}
 }
 
+func getAutocompleteData() *model.AutocompleteData {
+	jenkins := model.NewAutocompleteData("jenkins", "[subcommand]", "A Mattermost plugin to interact with Jenkins")
+
+	connect := model.NewAutocompleteData("connect", "[username] [APIToken]", "Connect to your Jenkins account")
```
Comment (mickmister) [inline server/command.go:59]:
```suggestion
	connect := model.NewAutocompleteData("connect", "[username] [APIToken]", "Connect your Mattermost account to your Jenkins account")
```

File: server/command.go [hunk 5]
```diff
@@ -49,9 +49,73 @@ func getCommand() *model.Command {
...
+	connect.AddTextArgument("Your username of Jenkins", "[username]", "")
+	connect.AddTextArgument("You API token from Jenkins", "[API token]", "")
+
+	disconnect := model.NewAutocompleteData("disconnect", "", "Disconnect from your Jenkins account")
```
Comment (mickmister) [inline server/command.go:63]:
```suggestion
	disconnect := model.NewAutocompleteData("disconnect", "", "Disconnect your Mattermost account from your Jenkins account")
```

File: server/command.go [hunk 6]
```diff
@@ -49,9 +49,73 @@ func getCommand() *model.Command {
...
+
+	createjob := model.NewAutocompleteData("createjob", "", "Create a Jenkins job using contents of config.xml")
+
+	build := model.NewAutocompleteData("build", "[jobname]", "Trigger a build for the giben job")
```
Comment (mickmister) [inline server/command.go:67]:
```suggestion
	build := model.NewAutocompleteData("build", "[jobname]", "Trigger a build for a given job")
```

File: server/command.go [hunk 7]
```diff
@@ -49,9 +49,73 @@ func getCommand() *model.Command {
...
+	build.AddTextArgument("folder1/jobname if the job is in a folder, or \"job with space\"", "[jobname]", "")
+
+	abort := model.NewAutocompleteData("abort", "[jobname] <build number>", "Abort the given build of the specified job")
+	abort.AddTextArgument("Job you want to abort a build to", "[jobname]", "")
```
Comment (mickmister) [inline server/command.go:71]:
```suggestion
	abort.AddTextArgument("Job associated with the build you want to abort", "[jobname]", "")
```

File: server/command.go [hunk 8]
```diff
@@ -49,9 +49,73 @@ func getCommand() *model.Command {
...
+	getArtifacts.AddTextArgument("The job you want to get artifacts from", "[jobname]", "")
+
+	testResults := model.NewAutocompleteData("test-results", "[jobname]", "Get test results of the last build of the given job")
+	getArtifacts.AddTextArgument("The job you want to get test results from", "[jobname]", "")
+
+	getLog := model.NewAutocompleteData("get-log", "[jobname] <build number>", "Get log of a build of the given job")
+	getLog.AddTextArgument("The job you want to get log from", "[jobname]", "")
+	getLog.AddTextArgument("Build number to get log from. If not specified, the last build is chosen", "<build number>", "")
+
+	plugins := model.NewAutocompleteData("plugins", "", "Get a list of installed plugins on Jenkins server")
```
Comment (mickmister) [inline server/command.go:93]:
```suggestion
	plugins := model.NewAutocompleteData("plugins", "", "Get a list of installed plugins on the Jenkins server")
```

File: server/command.go [hunk 9]
```diff
@@ -49,9 +49,73 @@ func getCommand() *model.Command {
...
+
+	safeRestart := model.NewAutocompleteData("safe-restart", "", "Safe restart of the Jenkins server")
+
+	me := model.NewAutocompleteData("me", "", "Display the connected Jenkins account")
+
+	help := model.NewAutocompleteData("help", "", "Find help related to the syntaxc of the slash commands")
```
Comment (mickmister) [inline server/command.go:99]:
```suggestion
	help := model.NewAutocompleteData("help", "", "Find help related to the syntax of the slash commands")
```

File: server/command.go [hunk 10]
```diff
@@ -49,9 +49,73 @@ func getCommand() *model.Command {
...
+	jenkins.AddCommand(enable)
+	jenkins.AddCommand(disable)
+	jenkins.AddCommand(delete)
+	jenkins.AddCommand(getArtifacts)
+	jenkins.AddCommand(testResults)
+	jenkins.AddCommand(getLog)
+	jenkins.AddCommand(plugins)
+	jenkins.AddCommand(safeRestart)
+	jenkins.AddCommand(me)
+	jenkins.AddCommand(help)
```
Comment (mickmister) [inline server/command.go:115]: Can these be added in alphabetical order, so they are listed that way in the UI?

**Review by mickmister — CHANGES_REQUESTED** [review]
Great work @nathanaelhoun! I have a few requests regarding grammar, and a request for alphabetizing the commands.

**Comment by nathanaelhoun** [comment]
Thanks @mickmister for the suggestions, I still need to improve my English!
I have adressed your requests :D

**Review by mickmister — APPROVED** [review]
LGTM :+1:  Thanks @nathanaelhoun!

**Review by iomodo — APPROVED** [review]
LGTM! Thanks for the contribution @nathanaelhoun !

**Review by DHaussermann — APPROVED** [review]
Tested and paseed
- Tested all options have been included in auto-complete and are functional 
- Tested that auto-complete options work on mobile
- No issues found.
LGTM!

Thanks @nathanaelhoun for adding this improvment!

**Pull request merged**